File: org.kevoree.modeling.cpp.microframework/src/main/java/org/kevoree/modeling/api/trace/TraceSequence.kt
package org.kevoree.modeling.api.trace

import org.kevoree.modeling.api.KMFContainer
import org.kevoree.modeling.api.KMFFactory
import org.kevoree.modeling.api.util.ActionType
import org.kevoree.modeling.api.json.Lexer
import org.kevoree.modeling.api.json.Type
import org.kevoree.modeling.api.util.ByteConverter

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 05/08/13
 * Time: 19:52
 */

public trait TraceSequence {

    var traces : java.util.ArrayList<org.kevoree.modeling.api.trace.ModelTrace>

    var factory : KMFFactory?

    fun populate(addtraces : List<org.kevoree.modeling.api.trace.ModelTrace>) : org.kevoree.modeling.api.trace.TraceSequence {
        traces.addAll(addtraces);
        return this;
    }

    fun populateFromString(addtracesTxt : String) : org.kevoree.modeling.api.trace.TraceSequence{
        return populateFromStream(ByteConverter.byteArrayInputStreamFromString(addtracesTxt))
    }

    fun populateFromStream(inputStream : java.io.InputStream) : org.kevoree.modeling.api.trace.TraceSequence {
        var lexer: Lexer = Lexer(inputStream)
        var currentToken = lexer.nextToken()
        if(currentToken.tokenType != Type.LEFT_BRACKET){
            throw Exception("Bad Format : expect [")
        }
        currentToken = lexer.nextToken()
        val keys = java.util.HashMap<String,String>();
        var previousName : String? = null
        while (currentToken.tokenType != Type.EOF && currentToken.tokenType != Type.RIGHT_BRACKET) {
            if(currentToken.tokenType == Type.LEFT_BRACE){
                keys.clear();
            }
            if(currentToken.tokenType == Type.VALUE){
                if(previousName!= null){
                    keys.put(previousName!!,currentToken.value.toString());
                    previousName = null
                }   else {
                    previousName = currentToken.value.toString()
                }
            }

            if(currentToken.tokenType == Type.RIGHT_BRACE){
                when(keys.get("traceType")!!) {
                    ActionType.SET.toString() -> {
                        traces.add(ModelSetTrace(keys.get("src")!!,keys.get("refname")!!,keys.get("objpath"),keys.get("content"),keys.get("typename")));
                    }
                    ActionType.ADD.toString() -> {
                        traces.add(ModelAddTrace(keys.get("src")!!,keys.get("refname")!!,keys.get("previouspath")!!,keys.get("typename")));
                    }
                    ActionType.ADD_ALL.toString() -> {
                        traces.add(ModelAddAllTrace(keys.get("src")!!,keys.get("refname")!!,keys.get("content")?.split(";")?.toList(),keys.get("typename")?.split(";")?.toList()));
                    }
                    ActionType.REMOVE.toString() -> {
                        traces.add(ModelRemoveTrace(keys.get("src")!!,keys.get("refname")!!,keys.get("objpath")!!));
                    }
                    ActionType.REMOVE_ALL.toString() -> {
                        traces.add(ModelRemoveAllTrace(keys.get("src")!!,keys.get("refname")!!));
                    }
                    ActionType.RENEW_INDEX.toString() -> {
                    }
                    else -> {
                        println("Trace lost !!!" )
                    }
                }
            }
            currentToken = lexer.nextToken()
        }
        return this;
    }

    fun exportToString() : String {
        val buffer = StringBuilder()
        buffer.append("[")
        var isFirst = true
        for(trace in traces){
            if(!isFirst){
                buffer.append(",")
            }
            buffer.append(trace.toString())
            isFirst = false
        }
        buffer.append("]")
        return buffer.toString()
    }

    fun applyOn(target : org.kevoree.modeling.api.KMFContainer) : Boolean {
        val traceApplicator = org.kevoree.modeling.api.trace.ModelTraceApplicator(target,factory!!)
        traceApplicator.applyTraceOnModel(this)
        //TODO implements the result
        return true
    }
}

File: org.kevoree.modeling.cpp.microframework/src/main/cpp/org/kevoree/modeling/api/trace/TraceSequence.cpp
#include "TraceSequence.h"

#include <map>
#include <list>
#include <sstream>
#include "../json/Lexer.h"
#include "../utils/Hashmap.h"
#include <stdlib.h>
using namespace std;



TraceSequence::TraceSequence(){

}
	
TraceSequence::~TraceSequence(){
		for (std::list<ModelTrace*>::iterator iterator = traces.begin(), end = traces.end(); iterator != end; ++iterator) 
        {
			   ModelTrace * t = *iterator;
			   delete t;
		}	
		traces.clear();
}


TraceSequence* TraceSequence::populate(std::list<ModelTrace*> addtraces){  
	     std::copy(addtraces.begin(), addtraces.end(), std::back_insert_iterator<std::list<ModelTrace*> >(traces)); // addAll
        return this;	
}

    
void TraceSequence::append(TraceSequence seq){
	 std::copy(seq.traces.begin(), seq.traces.end(), std::back_insert_iterator<std::list<ModelTrace*> >(traces)); // addAll
}



TraceSequence* TraceSequence::populateFromString(string addtracesTxt){
		 std::istringstream ss( addtracesTxt); // convert string to istream
		 return populateFromStream(ss);	
}


TraceSequence* TraceSequence::populateFromStream(istream &inputStream )
{
	  Lexer *lexer= new Lexer(inputStream);
      Token currentToken = lexer->nextToken();
  
        if(currentToken.tokenType != LEFT_BRACKET){
            throw ("Bad Format : expect [");
        }
        
      currentToken = lexer->nextToken();
      Hashmap<string> *keys = new Hashmap<string>();
      string previousName;
      ModelTrace *modeltrace;
      
      while (currentToken.tokenType != END_OF_FILE && currentToken.tokenType != RIGHT_BRACKET) {
				
	      if(currentToken.tokenType == LEFT_BRACE){
				delete keys;
				keys = new Hashmap<string>();
          }
          
          if(currentToken.tokenType == VALUE){
                if(!previousName.empty()){
                    keys->insert(previousName,currentToken.value);
                    previousName.clear();
                }   else {
                    previousName = currentToken.value;
                }
            }
            
             if(currentToken.tokenType == RIGHT_BRACE){
			
					if(keys->find("traceType") != 0){
						
						string traceType = *(keys->find("traceType"));
						string _src = "";
						string _refname = "";
						string _objpath = "";
						string _content = "";
						string _typename = "";
						string _previouspath = "";
						list < string > _l_previousPath;
						list < string > _l_typeName;
								
						switch(atoi(traceType.c_str())){
							
							case SET:
					
								if(keys->find("src") != 0){
									_src =*(keys->find("src"));
								}
								if(keys->find("src") != 0){
									_src =*(keys->find("src"));
								}
								if(keys->find("refname") != 0){
									_refname =*(keys->find("refname"));
								}		
								if(keys->find("objpath") != 0){
									_objpath =*(keys->find("objpath"));
								}	
								if(keys->find("content") != 0){
									_content =*(keys->find("content"));
								}
								if(keys->find("typename") != 0){
									_typename =*(keys->find("typename"));
								}	
								modeltrace = new ModelSetTrace(_src,_refname,_objpath,_content,_typename);
		
								traces.push_back(modeltrace);
			
							break;
							
							case ADD:
								if(keys->find("src") != 0){
									_src =*(keys->find("src"));
								}
								if(keys->find("refname") != 0){
									_refname =*(keys->find("refname"));
								}	
								if(keys->find("previouspath") != 0){
									_previouspath =*(keys->find("previouspath"));
								}	
								if(keys->find("typename") != 0){
									_typename =*(keys->find("typename"));
								}
								modeltrace =new ModelAddTrace(_src,_refname,_previouspath,_typename);
								traces.push_back(modeltrace);					
							break;
							
							case REMOVE:
							
								if(keys->find("src") != 0){
									_src =*(keys->find("src"));
								}
								if(keys->find("refname") != 0){
									_refname =*(keys->find("refname"));
								}		
								if(keys->find("objpath") != 0){
									_objpath =*(keys->find("objpath"));
								}	
								modeltrace= new ModelRemoveTrace(_src,_refname,_objpath);
								traces.push_back(modeltrace);
							break;
							
							case ADD_ALL:
							/*
								if(keys->find("src") != 0){
									_src =*(keys->find("src"));
								}
								if(keys->find("refname") != 0){
									_refname =*(keys->find("refname"));
								}	
								if(keys->find("content") != 0){
									_content =*(keys->find("content"));
								}	
								*/
										cout << "TODO ADD_ALL !!!" <<endl;
							break;
							
							case REMOVE_ALL:
								cout << "TODO REMOVE_ALL !!!" <<endl;
							
							break;
							
							case RENEW_INDEX:
									cout << "TODO RENEW_INDEX !!!" <<endl;
							break;
							
							default :
								cout << "TRACE LOST !!!" <<endl;
							break;							
						}
					}
			 }			
               currentToken = lexer->nextToken(); 
    }
    delete keys;
    return this;        
}

string TraceSequence::exportToString(){
		string buffer;
		buffer.append("[");
        bool isFirst = true;
     	for (std::list<ModelTrace*>::const_iterator iterator = traces.begin(), end = traces.end(); iterator != end; ++iterator) 
        {
			if(!isFirst){
                buffer.append(",");
            }
			ModelTrace *t = *iterator;	
			buffer.append(t->toString());
            isFirst = false;
		}
        buffer.append("]");
        return buffer;
    }
    
    
string TraceSequence::toString()  {
   return exportToString();
}

void TraceSequence::reverse(){
		traces.reverse();
}
